scripts: west_commands: blobs: support blobs cache

thorsten-klein opened the pull request:
This PR adds support for cache directories for blobs during `west blobs fetch` via the west config option `blobs.cache-dirs` or cli argument `--cache-dirs`. The variable may contain a single directory or multiple directories separated by `;`.

Additionally, via `blobs.auto-cache` config option or the `--auto-cache` cli argument an auto-cache directory can be specified. Thi auto-cache is automatically filled when a blob is downloaded, whereby the filename in cache is suffixed by the blob's SHA-256 hash (<filename>.<sha>).

`west blobs fetch` will search in all configured cache directories (including the auto-cache). 
The file in the cache directories can be named either as the blob file (e.g. `some_blob.bin.se`) or suffixed with its hashsum (e.g. `some_blob.bin.se.b94c34fc2a638a1713c79b04992020241aad32bab2be4a85623c8b5bb715e36e`)
If the blob is found, it is copied from the cache to the blobs path; otherwise, it is downloaded from the URL as usual.


### Why?

This feature can help to speed up builds and save network bandwidth in a CI environment, as blobs are cached and not re-downloaded. 
It also helps to become independent from 3rd party servers, as blobs can new be pre-fetched, packaged and uploaded to own server. Later, they can be downloaded and unpacked (to one or more directories) and provided as cache directories to `west blobs fetch`.

### Considered Alternative

Another considered solution was the support for some remapping of urls from blobs, e.g. via config file or manifest file. But I have discarded this idea as the solution from this PR is much simpler.

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -121,16 +123,51 @@ def list(self, args):
         inst = fetcher()
+
+        # Fill the cache if only one cache dir is used.
+        if not cached_blob and len(cache_dirs) == 1:
```
This feels like a hidden feature, either we need a separate option, or we need to split into two options (similar to `auto-cache` for west)

thorsten-klein replied:
Any proposal for the west config option? Maybe `blobs.auto-cache-dir` which may always contain only one path?

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -121,16 +131,58 @@ def list(self, args):
+        auto_cache_dir = args.auto_cache or config.get('blobs', 'auto-cache', fallback=None)
+
+        # convert all cache dirs to Path
+        cache_paths = [Path(p) for p in cache_dirs.split(';') if p]
```
No need to check if the paths exist?

marc-hb commented on scripts/west_commands/blobs.py:
```diff
@@ -5,12 +5,14 @@
 from urllib.parse import urlparse
 
 from west.commands import WestCommand
+from west.configuration import config
```
```suggestion
from west.configuration import wconfig
```

`config` is incredibly vague. It is a variable, a function, a module, an object?

marc-hb commented on scripts/west_commands/blobs.py:
```diff
@@ -121,16 +131,58 @@ def list(self, args):
+        cache_dirs = args.cache_dirs or config.get('blobs', 'cache-dirs', fallback='')
+        auto_cache_dir = args.auto_cache or config.get('blobs', 'auto-cache', fallback=None)
+
+        # convert all cache dirs to Path
```
```suggestion
```
Comments should add information or structure, not paraphrase obvious one-liners.

marc-hb commented on scripts/west_commands/blobs.py:
```diff
@@ -121,16 +131,58 @@ def list(self, args):
         inst = fetcher()
+
+        # Fill the auto-cache if in use
+        if not cached_blob and auto_cache_dir:
```
Why keep printing "Found fetcher" in the debug logs when it won't be used at all?

I agree with SonarQubeCloud that this function is getting a bit long; you can probably split it in two: 1. `get_blob()` that searches the cache and invokes `fetch_blob()` when not found. Also, probably reserve the term "fetch" for the network and use a different one for the cache to make the distinction clearer.

thorsten-klein committed:
scripts: west_commands: format blobs.py

blobs.py is formatted with `ruff format`

Signed-off-by: Thorsten Klein <Thorsten.Klein@bshg.com>

thorsten-klein committed:
scripts: west_commands: blobs: support blobs cache

An auto-cache directory can be provided via the `west blobs fetch
--auto-cache` argument or the `blobs.auto-cache-dir` config option.
When enabled, the auto-cache is automatically filled whenever a blob is
missing and must be downloaded.

One or more additional cache directories can be specified via argument
`west blobs fetch --cache-dirs` or the `blobs.cache-dir` config option
(multiple paths separated by `;`).

`west blobs fetch` searches all configured cache directories (including
the auto-cache) for a matching blob filename. Cached files may be stored
either under their original filename or with a SHA-256 suffix
(`<filename>.<sha>`).
If found, the blob is copied from the cache to the blob path; otherwise
it is downloaded from its url to the blob path.

Signed-off-by: Thorsten Klein <Thorsten.Klein@bshg.com>

thorsten-klein committed:
doc: develop: west: add documentation for west blobs caches

Documentation for west blobs `--cache-dirs` and `--auto-cache`,
respectively config options `blobs.cache-dirs` and `blobs.auto-cache`.

Signed-off-by: Thorsten Klein <Thorsten.Klein@bshg.com>

thorsten-klein committed:
.ruff-excludes: remove scripts/west_commands/blobs.py

Regarding documentation, when changing an excluded file, contributors
are encouraged to remove it from the list and format it in a separate
commit.

Signed-off-by: Thorsten Klein <Thorsten.Klein@bshg.com>

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -100,6 +101,18 @@ def do_add_parser(self, parser_adder):
+            '--cache-dirs',
+            help='''Semicolon-separated list of directories to search for cached
+                                    blobs before downloading. Cache files may use the original
+                                    filename or be suffixed with `.<sha256>`.''',
```
`help` indentation looks awkward.

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -61,22 +62,30 @@ def do_add_parser(self, parser_adder):
+            '-f',
+            '--format',
+            help='''format string to use to list each blob;
+                                    see FORMAT STRINGS below''',
```
Ditto

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -121,26 +146,107 @@ def list(self, args):
+            if not cache_dirs:
+                cache_dirs = self.config.get('blobs.cache-dirs')
+            if not auto_cache_dir:
+                auto_cache_dir = self.config.get('blobs.auto-cache')
```
If the user passed an empty string, it should take precedence over the config IMO, so comparing to `None`.

```suggestion
            if cache_dirs is None:
                cache_dirs = self.config.get('blobs.cache-dirs')
            if auto_cache_dir is None:
                auto_cache_dir = self.config.get('blobs.auto-cache')
```

pdgendt commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -220,6 +220,19 @@ the specific blobs that are fetched, by passing a regular expression::
   west blobs fetch hal_espressif --allow-regex 'lib/esp32/.*'
 
+An auto-cache directory can be provided via the ``--auto-cache`` cli argument
+or via the ``blobs.auto-cache-dir`` config option. When enabled, the auto-cache
```
Different name in the implementation

```suggestion
or via the ``blobs.auto-cache`` config option. When enabled, the auto-cache
```

carlescufi reviewed (approved):
Thanks for this! Checked the code and looks reasonable to me
